# HDB Resale House Price Prediction

## Preface

The issue of home affordability in Singapore has become increasingly prevalent in recent years. As highlighted in a recent article published by [The Straits Time](https://www.straitstimes.com/singapore/housing/hdb-resale-prices-rise-09-in-q1-smallest-increase-in-last-10-quarters#:~:text=In%20recent%20years%2C%20the%20issue,the%20private%20residential%20property%20market.), the resale prices of Housing and Development Board (HDB) flats have risen significantly by 31.9 percent over the past 12 quarters, surpassing the 27.9 percent increase observed in the private residential property market during the same period. The rise in resale flat prices in Singapore can be attributed to several factors: 
1. Limited supply: The supply of public housing in Singapore is limited compared to the high demand. The government's efforts to control the population growth and maintain a sustainable housing market have led to a controlled supply of new flats. This scarcity of supply puts upward pressure on resale flat prices. [Link](https://www.channelnewsasia.com/commentary/hdb-public-housing-bto-affordability-accessibility-supply-demand-3259186)
2. Population growth and immigration: Singapore's population growth grew from 4.03 million in 2000 to 5.7 million in 2019. A 41% increase over 20 years. The increase in population creates higher demand for housing, including resale flats, leading to price appreciation. [Link](https://www.straitstimes.com/singapore/curb-population-growth-to-reduce-housing-demand-need-to-clear-forests-leong-mun-wai)

3. Land scarcity: Singapore has limited land resources, and finding suitable land for new housing developments can be challenging. As a result, land prices are high, and the cost of construction and land acquisition is reflected in the resale flat prices. 
4. Housing policies: The Singapore government has implemented various policies to promote home ownership and provide affordable housing options. However, some of these policies, such as the Central Provident Fund (CPF) housing grants and restrictions on the purchase of new flats by certain groups, can contribute to increased demand for resale flats, thereby pushing up prices. 
5. Economic growth and prosperity: Singapore's strong economy and high standard of living contribute to the affordability of housing. As people become more affluent, they may be willing to pay higher prices for better quality and well-located resale flats. 
6. Renovation and upgrading: Many resale flat buyers prefer flats that have undergone renovation or upgrading. The cost of these renovations and improvements, along with the added value they bring, can contribute to higher resale flat prices. 
7. Speculation and investment: Some buyers purchase resale flats as an investment, hoping to benefit from potential capital appreciation over time. Speculative activities can drive up prices in the resale flat market. 
8. Inflation and construction costs: Inflation and rising construction costs can also contribute to the increase in resale flat prices. As the cost of construction materials, labor, and land acquisition rises, developers and sellers may adjust their prices accordingly. 

Furthermore, the escalating costs and lengthy waiting [time](https://www.straitstimes.com/singapore/housing/bto-flat-waiting-times-between-4-and-5-years-even-with-covid-19-delays-desmond-lee) for Build-To-Order (BTO) flats have dissuaded numerous newlywed couples from purchasing them, leading to a surge in demand for HDB resale flats. Against this backdrop, the prediction of HDB resale house prices serves as a valuable tool to assist both buyers and sellers in making informed decisions.

## Project Stakeholder

1. Singapore public who need information related to resale HDB flats from 1990 to 2023.
2. Singaporean/permanent residence/foreigner who wants to buy resale HDB flats and needs early information related to resale HDB flats prices from 1990 to 2023.
3. Anyone who has interested in resale HDB flats prices from 1990 to 2023.
4. Any sellers that wants to check on their current HDB flat prices from 2023 onwards to sell.
5. Any buyers that wants to check on their current HDB flat prices from 2023 onwards to buy.

## Objective

The objective is to provide a clear and concise overview of the issue of home affordability in Singapore, particularly with regards to the rising resale prices of HDB flats and the challenges faced by newlywed couples in purchasing a home. This project aims to establish the significance of predicting HDB resale house prices as a means to aid both buyers and sellers in making informed decisions.

## Datasets

The dataset has been taken from Government official site. [Link](https://data.gov.sg/dataset/resale-flat-prices)

## Importing relevant libraries

In [1]:
#import libries
import numpy as np
import pandas as pd
import re

In [2]:
df1 = pd.read_csv('./datasets/resale-flat-prices-based-on-approval-date-1990-1999.csv', low_memory = False)

In [3]:
df2 = pd.read_csv('./datasets/resale-flat-prices-based-on-approval-date-2000-feb-2012.csv', low_memory = False)

In [4]:
df3 = pd.read_csv('./datasets/resale-flat-prices-based-on-registration-date-from-jan-2015-to-dec-2016.csv', low_memory = False)

In [5]:
df4 = pd.read_csv('./datasets/resale-flat-prices-based-on-registration-date-from-jan-2017-onwards.csv', low_memory = False)

In [6]:
df5 = pd.read_csv('./datasets/resale-flat-prices-based-on-registration-date-from-mar-2012-to-dec-2014.csv', low_memory = False)

In [7]:
df1.head()

month        town flat_type block       street_name storey_range  \
0  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
1  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     04 TO 06   
2  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
3  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     07 TO 09   
4  1990-01  ANG MO KIO    3 ROOM   216  ANG MO KIO AVE 1     04 TO 06   

   floor_area_sqm      flat_model  lease_commence_date  resale_price  
0            31.0        IMPROVED                 1977          9000  
1            31.0        IMPROVED                 1977          6000  
2            31.0        IMPROVED                 1977          8000  
3            31.0        IMPROVED                 1977          6000  
4            73.0  NEW GENERATION                 1976         47200

In [8]:
df2.head()

month        town flat_type block       street_name storey_range  \
0  2000-01  ANG MO KIO    3 ROOM   170  ANG MO KIO AVE 4     07 TO 09   
1  2000-01  ANG MO KIO    3 ROOM   174  ANG MO KIO AVE 4     04 TO 06   
2  2000-01  ANG MO KIO    3 ROOM   216  ANG MO KIO AVE 1     07 TO 09   
3  2000-01  ANG MO KIO    3 ROOM   215  ANG MO KIO AVE 1     07 TO 09   
4  2000-01  ANG MO KIO    3 ROOM   218  ANG MO KIO AVE 1     07 TO 09   

   floor_area_sqm      flat_model  lease_commence_date  resale_price  
0            69.0        Improved                 1986      147000.0  
1            61.0        Improved                 1986      144000.0  
2            73.0  New Generation                 1976      159000.0  
3            73.0  New Generation                 1976      167000.0  
4            67.0  New Generation                 1976      163000.0

In [9]:
df3.head()

month        town flat_type block        street_name storey_range  \
0  2015-01  ANG MO KIO    3 ROOM   174   ANG MO KIO AVE 4     07 TO 09   
1  2015-01  ANG MO KIO    3 ROOM   541  ANG MO KIO AVE 10     01 TO 03   
2  2015-01  ANG MO KIO    3 ROOM   163   ANG MO KIO AVE 4     01 TO 03   
3  2015-01  ANG MO KIO    3 ROOM   446  ANG MO KIO AVE 10     01 TO 03   
4  2015-01  ANG MO KIO    3 ROOM   557  ANG MO KIO AVE 10     07 TO 09   

   floor_area_sqm      flat_model  lease_commence_date  remaining_lease  \
0            60.0        Improved                 1986               70   
1            68.0  New Generation                 1981               65   
2            69.0  New Generation                 1980               64   
3            68.0  New Generation                 1979               63   
4            68.0  New Generation                 1980               64   

   resale_price  
0      255000.0  
1      275000.0  
2      285000.0  
3      290000.0  
4      290000.0

In [10]:
df4.head()

month        town flat_type block        street_name storey_range  \
0  2017-01  ANG MO KIO    2 ROOM   406  ANG MO KIO AVE 10     10 TO 12   
1  2017-01  ANG MO KIO    3 ROOM   108   ANG MO KIO AVE 4     01 TO 03   
2  2017-01  ANG MO KIO    3 ROOM   602   ANG MO KIO AVE 5     01 TO 03   
3  2017-01  ANG MO KIO    3 ROOM   465  ANG MO KIO AVE 10     04 TO 06   
4  2017-01  ANG MO KIO    3 ROOM   601   ANG MO KIO AVE 5     01 TO 03   

   floor_area_sqm      flat_model  lease_commence_date     remaining_lease  \
0            44.0        Improved                 1979  61 years 04 months   
1            67.0  New Generation                 1978  60 years 07 months   
2            67.0  New Generation                 1980  62 years 05 months   
3            68.0  New Generation                 1980   62 years 01 month   
4            67.0  New Generation                 1980  62 years 05 months   

   resale_price  
0      232000.0  
1      250000.0  
2      262000.0  
3      265000.0  
4      265000.0

In [11]:
df5.head()

month        town flat_type block        street_name storey_range  \
0  2012-03  ANG MO KIO    2 ROOM   172   ANG MO KIO AVE 4     06 TO 10   
1  2012-03  ANG MO KIO    2 ROOM   510   ANG MO KIO AVE 8     01 TO 05   
2  2012-03  ANG MO KIO    3 ROOM   610   ANG MO KIO AVE 4     06 TO 10   
3  2012-03  ANG MO KIO    3 ROOM   474  ANG MO KIO AVE 10     01 TO 05   
4  2012-03  ANG MO KIO    3 ROOM   604   ANG MO KIO AVE 5     06 TO 10   

   floor_area_sqm      flat_model  lease_commence_date  resale_price  
0            45.0        Improved                 1986      250000.0  
1            44.0        Improved                 1980      265000.0  
2            68.0  New Generation                 1980      315000.0  
3            67.0  New Generation                 1984      320000.0  
4            67.0  New Generation                 1980      321000.0

## Data Cleaning

1. To Combine all dataframes together to form a singular dataframe.

In [12]:
df = pd.concat([df1,df2, df3, df4, df5])

In [13]:
df.head()

month        town flat_type block       street_name storey_range  \
0  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
1  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     04 TO 06   
2  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
3  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     07 TO 09   
4  1990-01  ANG MO KIO    3 ROOM   216  ANG MO KIO AVE 1     04 TO 06   

   floor_area_sqm      flat_model  lease_commence_date  resale_price  \
0            31.0        IMPROVED                 1977        9000.0   
1            31.0        IMPROVED                 1977        6000.0   
2            31.0        IMPROVED                 1977        8000.0   
3            31.0        IMPROVED                 1977        6000.0   
4            73.0  NEW GENERATION                 1976       47200.0   

  remaining_lease  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN

In [14]:
df.shape

(899002, 11)

Remarks: The various dataframes have been combined together.

The function below is to show a summary of information for df.

In [15]:
# summary creation
def summary(df):
    #show all columns
    pd.set_option('display.max_rows', None)
    
    #shows number of columns and rows
    shape = df.shape
    print('='*80)
    print('Shape of dataframe')
    print(shape)
    
    #checking for null values
    null_val = df.isnull().sum().sort_values(ascending=False)
    print('='*80)
    indicator = False
    print('Missing values distribution (%):')
    null_val = df.isnull().mean().sort_values(ascending=False)*100
    for i in range(len(null_val)):
        if (null_val[i] != 0):
            print(f'{null_val.index[i]}: {null_val[i]:.4f}')
            indicator = True
        if null_val[i] == 0:
            continue
    if indicator == False:
        print('no null values (ﾉ^_^)ﾉ')
    print('='*80)
    # print('Columns with duplicate values')
    # duplicated = df[df.duplicated()]
    # if duplicated.size > 0:
    #     print (duplicated)
    # else: 
    #     print('No duplicated rows ๑(◕‿◕)๑')
    # print('='*80)
    
    #checking datatype in each column
    print('Column datatypes: ')
    print(df.dtypes)

In [16]:
summary(df)

Shape of dataframe
(899002, 11)
Missing values distribution (%):
remaining_lease: 78.8708
Column datatypes: 
month                   object
town                    object
flat_type               object
block                   object
street_name             object
storey_range            object
floor_area_sqm         float64
flat_model              object
lease_commence_date      int64
resale_price           float64
remaining_lease         object
dtype: object


remarks: function summary created.

Remarks:
1. There are 899,002 rows and 11 columns
2. There are 78.8% of data missing in column 'remaining lease'
3. There are some column datatypes being classified wrongly

The subsequent part of this notebook is to address such problems

Converting month to datetime format

In [17]:
df['month'] = pd.to_datetime(df['month'])

Remarks: date has been converted to correct datetime format.

Political parties such as PAP/WP/PSP etc. could play a part in affecting HDB house prices. Hence, political parties will be featured engineered to account for possibly influencing the house prices.

In [18]:
df['political_parties'] = 'RULING'

In [19]:
df.loc[df['town'] == 'HOUGANG', 'political_parties'] = 'OPPOSITION'

In [20]:
df.loc[(df['town'] == 'SENGKANG') & (df['month'].dt.year >= 2020), 'political_parties'] = 'OPPOSITION'

In [21]:
df.loc[(df['town'] == 'SERANGON') & (df['month'].dt.year >= 2010), 'political_parties'] = 'OPPOSITION'

### Checking each individual columns

In [22]:
df['town'].unique()

array(['ANG MO KIO', 'BEDOK', 'BISHAN', 'BUKIT BATOK', 'BUKIT MERAH',
       'BUKIT TIMAH', 'CENTRAL AREA', 'CHOA CHU KANG', 'CLEMENTI',
       'GEYLANG', 'HOUGANG', 'JURONG EAST', 'JURONG WEST',
       'KALLANG/WHAMPOA', 'MARINE PARADE', 'QUEENSTOWN', 'SENGKANG',
       'SERANGOON', 'TAMPINES', 'TOA PAYOH', 'WOODLANDS', 'YISHUN',
       'LIM CHU KANG', 'SEMBAWANG', 'BUKIT PANJANG', 'PASIR RIS',
       'PUNGGOL'], dtype=object)

In [23]:
df['flat_type'].unique()

array(['1 ROOM', '3 ROOM', '4 ROOM', '5 ROOM', '2 ROOM', 'EXECUTIVE',
       'MULTI GENERATION', 'MULTI-GENERATION'], dtype=object)

In [24]:
df['flat_model'].unique()

array(['IMPROVED', 'NEW GENERATION', 'MODEL A', 'STANDARD', 'SIMPLIFIED',
       'MODEL A-MAISONETTE', 'APARTMENT', 'MAISONETTE', 'TERRACE',
       '2-ROOM', 'IMPROVED-MAISONETTE', 'MULTI GENERATION',
       'PREMIUM APARTMENT', 'Improved', 'New Generation', 'Model A',
       'Standard', 'Apartment', 'Simplified', 'Model A-Maisonette',
       'Maisonette', 'Multi Generation', 'Adjoined flat',
       'Premium Apartment', 'Terrace', 'Improved-Maisonette',
       'Premium Maisonette', '2-room', 'Model A2', 'Type S1', 'Type S2',
       'DBSS', 'Premium Apartment Loft', '3Gen'], dtype=object)

In [25]:
list(df['block'].unique())

['309',
 '216',
 '211',
 '202',
 '235',
 '232',
 '308',
 '220',
 '219',
 '247',
 '320',
 '252',
 '223',
 '230',
 '329',
 '313',
 '117',
 '110',
 '343',
 '345',
 '346',
 '121',
 '129',
 '130',
 '128',
 '127',
 '126',
 '403',
 '404',
 '405',
 '417',
 '418',
 '419',
 '441',
 '442',
 '443',
 '444',
 '450',
 '435',
 '433',
 '434',
 '424',
 '425',
 '466',
 '471',
 '474',
 '570',
 '586',
 '455',
 '465',
 '463',
 '564',
 '560',
 '558',
 '559',
 '538',
 '534',
 '601',
 '603',
 '604',
 '608',
 '611',
 '505',
 '503',
 '610',
 '607',
 '524',
 '513',
 '643',
 '542',
 '548',
 '550',
 '639',
 '637',
 '330',
 '333',
 '156',
 '152',
 '178',
 '180',
 '209',
 '231',
 '254',
 '103',
 '105',
 '344',
 '324',
 '120',
 '124',
 '414',
 '438',
 '427',
 '428',
 '473',
 '573',
 '585',
 '456',
 '544',
 '640',
 '638',
 '646',
 '150',
 '179',
 '336',
 '335',
 '401',
 '439',
 '430',
 '460',
 '459',
 '716',
 '545',
 '620',
 '622',
 '259',
 '101',
 '18',
 '28',
 '29',
 '30',
 '75',
 '76',
 '501',
 '502',
 '504',
 '2',


Remarks: 
1. 'Block' column could be featured engineered to split up the alphabets from the numbers, so as to reduce dimentionality.

### Checking for null values

Observations: 
1. It seems that there are only values for remaining_lease from 2015 to 2023. 
2. The 76.8% of data missing in colunmn remaining_lease shows that there are not data collected for it before 2015.

Strategy: 
1. HDB flats are sold on a 99 year lease. The remaining_lease column is simply a subtraction of HDB lease_commence_date to month recorded. 
2. Feature engineer a new column to recalculate the remaining_lease year.

In [26]:
df.columns

Index(['month', 'town', 'flat_type', 'block', 'street_name', 'storey_range',
       'floor_area_sqm', 'flat_model', 'lease_commence_date', 'resale_price',
       'remaining_lease', 'political_parties'],
      dtype='object')

In [27]:
df['lease_commence_date'].head(10)

0    1977
1    1977
2    1977
3    1977
4    1976
5    1977
6    1977
7    1977
8    1977
9    1977
Name: lease_commence_date, dtype: int64

In [28]:
df['lease_commence_date'] = df['lease_commence_date'].apply(lambda x: str(x))

In [29]:
df['lease_commence_date'] = pd.to_datetime(df['lease_commence_date'], format='%Y')

In [30]:
df['lease_commence_date'].head(10)

0   1977-01-01
1   1977-01-01
2   1977-01-01
3   1977-01-01
4   1976-01-01
5   1977-01-01
6   1977-01-01
7   1977-01-01
8   1977-01-01
9   1977-01-01
Name: lease_commence_date, dtype: datetime64[ns]

In [31]:
df['lease_commence_date'].dtype

dtype('<M8[ns]')

In [32]:
df['end_of_lease'] = df['lease_commence_date'] + pd.DateOffset(years=99)

In [33]:
df['end_of_lease'] = pd.to_datetime(df['end_of_lease']).dt.date

In [34]:
df['end_of_lease'] = pd.to_datetime(df['end_of_lease'])

In [35]:
df['end_of_lease'].head()

0   2076-01-01
1   2076-01-01
2   2076-01-01
3   2076-01-01
4   2075-01-01
Name: end_of_lease, dtype: datetime64[ns]

In [36]:
df['month'].head()

0   1990-01-01
1   1990-01-01
2   1990-01-01
3   1990-01-01
4   1990-01-01
Name: month, dtype: datetime64[ns]

In [37]:
# df['remaining_lease_days'] = (df['end_of_lease'] - df['month']).dt.days

In [38]:
# df['remaining_lease_days'].head()

In [39]:
# df['remaining_lease_days'].dtype

In [40]:
df.head()

month        town flat_type block       street_name storey_range  \
0 1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
1 1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     04 TO 06   
2 1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
3 1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     07 TO 09   
4 1990-01-01  ANG MO KIO    3 ROOM   216  ANG MO KIO AVE 1     04 TO 06   

   floor_area_sqm      flat_model lease_commence_date  resale_price  \
0            31.0        IMPROVED          1977-01-01        9000.0   
1            31.0        IMPROVED          1977-01-01        6000.0   
2            31.0        IMPROVED          1977-01-01        8000.0   
3            31.0        IMPROVED          1977-01-01        6000.0   
4            73.0  NEW GENERATION          1976-01-01       47200.0   

  remaining_lease political_parties end_of_lease  
0             NaN            RULING   2076-01-01  
1             NaN            RULING   2076-01-01  
2             NaN            RULING   2076-01-01  
3             NaN            RULING   2076-01-01  
4             NaN            RULING   2075-01-01

Remarks: 
1. Date for end of lease is calculated
2. remainding lease days columns calculated

Drop unnecessary columns such to reduce dimensionality of data

In [41]:
df.drop(['remaining_lease'], axis = 1, inplace = True)

In [42]:
df.head(5)

month        town flat_type block       street_name storey_range  \
0 1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
1 1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     04 TO 06   
2 1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
3 1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     07 TO 09   
4 1990-01-01  ANG MO KIO    3 ROOM   216  ANG MO KIO AVE 1     04 TO 06   

   floor_area_sqm      flat_model lease_commence_date  resale_price  \
0            31.0        IMPROVED          1977-01-01        9000.0   
1            31.0        IMPROVED          1977-01-01        6000.0   
2            31.0        IMPROVED          1977-01-01        8000.0   
3            31.0        IMPROVED          1977-01-01        6000.0   
4            73.0  NEW GENERATION          1976-01-01       47200.0   

  political_parties end_of_lease  
0            RULING   2076-01-01  
1            RULING   2076-01-01  
2            RULING   2076-01-01  
3            RULING   2076-01-01  
4            RULING   2075-01-01

Remarks:  
1. Columns dropped

### Renaming columns for better readability

In [43]:
df.rename(columns={'month': 'transaction_date'}, inplace=True)

In [44]:
df.head()

transaction_date        town flat_type block       street_name storey_range  \
0       1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
1       1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     04 TO 06   
2       1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
3       1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     07 TO 09   
4       1990-01-01  ANG MO KIO    3 ROOM   216  ANG MO KIO AVE 1     04 TO 06   

   floor_area_sqm      flat_model lease_commence_date  resale_price  \
0            31.0        IMPROVED          1977-01-01        9000.0   
1            31.0        IMPROVED          1977-01-01        6000.0   
2            31.0        IMPROVED          1977-01-01        8000.0   
3            31.0        IMPROVED          1977-01-01        6000.0   
4            73.0  NEW GENERATION          1976-01-01       47200.0   

  political_parties end_of_lease  
0            RULING   2076-01-01  
1            RULING   2076-01-01  
2            RULING   2076-01-01  
3            RULING   2076-01-01  
4            RULING   2075-01-01

Remarks:
1. Columns renamed.

In [45]:
df['flat_model'].unique()

array(['IMPROVED', 'NEW GENERATION', 'MODEL A', 'STANDARD', 'SIMPLIFIED',
       'MODEL A-MAISONETTE', 'APARTMENT', 'MAISONETTE', 'TERRACE',
       '2-ROOM', 'IMPROVED-MAISONETTE', 'MULTI GENERATION',
       'PREMIUM APARTMENT', 'Improved', 'New Generation', 'Model A',
       'Standard', 'Apartment', 'Simplified', 'Model A-Maisonette',
       'Maisonette', 'Multi Generation', 'Adjoined flat',
       'Premium Apartment', 'Terrace', 'Improved-Maisonette',
       'Premium Maisonette', '2-room', 'Model A2', 'Type S1', 'Type S2',
       'DBSS', 'Premium Apartment Loft', '3Gen'], dtype=object)

Observation:
1. It seems that there are some houses that are classified as the same, just that they are written in different manner. Example: 'MODEL A' and 'Model A'.

Strategy:
1. Renaming of items in the column.

In [46]:
df['flat_model'] = df['flat_model'].replace({'Improved':'IMPROVED',
                                            'New Generation': 'NEW GENERATION',
                                            'Model A': 'MODEL A',
                                            'Model A-Maisonette': 'MODEL A-MAISONETTE',
                                            'Standard':'STANDARD',
                                            'Apartment':'APARTMENT',
                                            'Multi Generation': 'MULTI GENERATION',
                                            'Premium Apartment':'PREMIUM APARTMENT',
                                            'Improved-Maisonette': 'IMPROVED-MAISONETTE',
                                            '2-room': '2-ROOM',
                                            'Terrace': 'TERRACE',
                                            '3Gen': 'MULTI GENERATION'})

In [47]:
df['flat_model'] = df['flat_model'].apply(lambda x: x.upper())

In [48]:
df['flat_model'].unique()

array(['IMPROVED', 'NEW GENERATION', 'MODEL A', 'STANDARD', 'SIMPLIFIED',
       'MODEL A-MAISONETTE', 'APARTMENT', 'MAISONETTE', 'TERRACE',
       '2-ROOM', 'IMPROVED-MAISONETTE', 'MULTI GENERATION',
       'PREMIUM APARTMENT', 'ADJOINED FLAT', 'PREMIUM MAISONETTE',
       'MODEL A2', 'TYPE S1', 'TYPE S2', 'DBSS', 'PREMIUM APARTMENT LOFT'],
      dtype=object)

In [49]:
df['flat_type'].unique()

array(['1 ROOM', '3 ROOM', '4 ROOM', '5 ROOM', '2 ROOM', 'EXECUTIVE',
       'MULTI GENERATION', 'MULTI-GENERATION'], dtype=object)

In [50]:
df['flat_type'] = df['flat_type'].replace({'MULTI-GENERATION':'MULTI GENERATION'})

Remarks:
1. Items has been renamed.

In [51]:
df.head()

transaction_date        town flat_type block       street_name storey_range  \
0       1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
1       1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     04 TO 06   
2       1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
3       1990-01-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     07 TO 09   
4       1990-01-01  ANG MO KIO    3 ROOM   216  ANG MO KIO AVE 1     04 TO 06   

   floor_area_sqm      flat_model lease_commence_date  resale_price  \
0            31.0        IMPROVED          1977-01-01        9000.0   
1            31.0        IMPROVED          1977-01-01        6000.0   
2            31.0        IMPROVED          1977-01-01        8000.0   
3            31.0        IMPROVED          1977-01-01        6000.0   
4            73.0  NEW GENERATION          1976-01-01       47200.0   

  political_parties end_of_lease  
0            RULING   2076-01-01  
1            RULING   2076-01-01  
2            RULING   2076-01-01  
3            RULING   2076-01-01  
4            RULING   2075-01-01

### Extracting out alphabets from block

As mentioned above, alphabets from block will be extracted to reduce dimensionality.

In [52]:
df['blk_num'] = df['block'].str.replace('[^0-9]', '')

C:\Users\johnk\AppData\Local\Temp\ipykernel_49208\705616691.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df['blk_num'] = df['block'].str.replace('[^0-9]', '')


In [53]:
df['blk_alpha'] = df['block'].str.extract('([A-Z]+)')

In [54]:
df['blk_alpha'] = df['blk_alpha'].apply(lambda x: '-' if pd.isna(x) else x)

In [55]:
df.drop(columns = 'block', inplace=True)

In [56]:
df.head()

transaction_date        town flat_type       street_name storey_range  \
0       1990-01-01  ANG MO KIO    1 ROOM  ANG MO KIO AVE 1     10 TO 12   
1       1990-01-01  ANG MO KIO    1 ROOM  ANG MO KIO AVE 1     04 TO 06   
2       1990-01-01  ANG MO KIO    1 ROOM  ANG MO KIO AVE 1     10 TO 12   
3       1990-01-01  ANG MO KIO    1 ROOM  ANG MO KIO AVE 1     07 TO 09   
4       1990-01-01  ANG MO KIO    3 ROOM  ANG MO KIO AVE 1     04 TO 06   

   floor_area_sqm      flat_model lease_commence_date  resale_price  \
0            31.0        IMPROVED          1977-01-01        9000.0   
1            31.0        IMPROVED          1977-01-01        6000.0   
2            31.0        IMPROVED          1977-01-01        8000.0   
3            31.0        IMPROVED          1977-01-01        6000.0   
4            73.0  NEW GENERATION          1976-01-01       47200.0   

  political_parties end_of_lease blk_num blk_alpha  
0            RULING   2076-01-01     309         -  
1            RULING   2076-01-01     309         -  
2            RULING   2076-01-01     309         -  
3            RULING   2076-01-01     309         -  
4            RULING   2075-01-01     216         -

In [57]:
df['blk_num'] = df['blk_num'].astype(int)

Remarks:
1. Columns created.

Upon inspection, it seems that street name will take up a lot of space when doing one hot encoding. Hence, the column will be dropped.

In [61]:
df.drop(columns = 'street_name', inplace=True)

Remarks:
1. Columns dropped.

In [62]:
df.to_csv('./datasets/cleaned_v1.csv', index=False)

## Summary

In [63]:
summary(df)

Shape of dataframe
(899002, 12)
Missing values distribution (%):
no null values (ﾉ^_^)ﾉ
Column datatypes: 
transaction_date       datetime64[ns]
town                           object
flat_type                      object
storey_range                   object
floor_area_sqm                float64
flat_model                     object
lease_commence_date    datetime64[ns]
resale_price                  float64
political_parties              object
end_of_lease           datetime64[ns]
blk_num                         int32
blk_alpha                      object
dtype: object


In summary:
1. There are 899,002 rows and 12 columns
2. There are no null values
3. The columns are in their correct data types